src/common/ui/codemirror-js-mixed/javascript-mixed.js

```diff
@@ -0,0 +1,810 @@
+        nextContext: null, // then exit local html mode
+        caseMatched: ctx => { ctx.style = exitLocalModeWithEndQuote(ctx.stream, ctx.state); },
+        caseNotMatched: ctx => { ctx.style = tokenInLocalModePlainString(ctx.stream, ctx.state); }, // else stay local mode
+      }),
+
+      // for HTML string template (without inline comment as a hint)
+      new Rule({
+        curContext: '<start>',
+        match: ctx => ctx.type === 'quasi' && RE_HTML_STRING_TEMPLATE.test(ctx.text),
+        nextContext: 'html-in',
+        caseMatched: ctx => prepReparseStringTemplateInLocalMode(htmlmixedMode,
+          ctx.stream, ctx.state),
+      }),
+
+      // for HTML string template (where first line is blank, html started in second line)
+      new Rule({
+        curContext: '<start>',
+        match: ctx => ctx.type === 'quasi' && /^[`](\\)?\s*$/.test(ctx.text), // first line is blank
+        nextContext: 'html-51',
+      }),
+      new Rule({
+        curContext: 'html-51',
+        match: ctx => ctx.type === 'quasi' && RE_HTML_BASE.test(ctx.text), // second line starts with a tag
+        nextContext: 'html-in',
+        caseMatched: ctx => prepReparseStringTemplateInLocalMode(htmlmixedMode,
+          ctx.stream, ctx.state, false),
+      }),
+
+    ];
+
+    // a map of all rules, keyed by curContext for quick look up during matching
+    const allRuleMap = (() => {
+      const res = {};
+      for (const rules of [htmlRules, cssRules]) {
+        for (const rule of rules) {
+          const key = rule.curContext;
+          res[key] = res[key] || [];
+          res[key].push(rule);
+        }
+      }
+      return res;
+    })();
+
+
+    function jsToken(stream, state) {
+      // dbg('jsToken -', `${stream.pos}: ${stream.string.substring(stream.pos).substring(0, 8)}`, state.lastType);
+
+      // adapt the existing jsMode tokenizer with the wrapper state
+      let tokenStyle = null;
+      if (!state.localMode) {
+        // when in local html/css context, skip js parsing,
+        // so as not to mess up js tokenizer's state.
+        tokenStyle = jsMode.token(stream, state.jsState);
+        dbg('jsMode.token - ', state.maybeLocalContext, state.jsState.lastType, stream.current(), `[${tokenStyle}]`);
+        if (tokenStyle === null) { // case the token is not relevant semantically, e.g., space or line break;
+          // just return,  skip local mode match,
+          // as such token is not reflected in stream/state so the local mode matcher
+          // will end up seeing previous token.
+          // dbg('   <--', `[${tokenStyle}]`, stream.current());
+          return null;
+        }
+      }
+
+      // optimization: short-circuit to skip local mode match when the rules won't cover
+      // Note: if the rules change (the <start> ones), the conditions here might need to be updated accordingly.
+      if (state.maybeLocalContext == null
+        && !['variable', 'comment', 'string', 'string-2'].includes(tokenStyle)) {
```

Comment:
I've tested a big script without html/css in strings at all and it opens only slightly slower than on the master (150 vs 133 which is expected), pastes just as fast at 90ms.

One thing that may influence the overall time covertly is garbage collector and here I thought this temporary array is probably created a lot of times. Plain comparison is less elegant but just as readable:

```js
      if (state.maybeLocalContext == null
      && tokenStyle !== 'variable'
      && tokenStyle !== 'comment'
      && tokenStyle !== 'string'
      && tokenStyle !== 'string-2') {
```

Reply:
hmm, I assume js engines could recognize such array as a constant and avoid creating them repeatedly. Better be safe than sorry. The latest commits include your suggestion, plus a similar fix where the codes returned `[localMode, localState]` array.

Concerning garbage collector impact, I had checked in a few tweaks to cut down object creation with that in mind: cut down `new RunContext()`, and avoid calling `stream.current()` (that creates new substring).